File: services/settlement/internal/products/products_test.go
package products

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"testing"
)

type tipleCase struct {
	Name            string   `json:"name"`
	Probabilities   []string `json:"probabilities"`
	MarginBp        int      `json:"marginBp"`
	OfferedOdds     string   `json:"offeredOdds"`
	FairProbability float64  `json:"fairProbability"`
}

type tippotCaseTier struct {
	K          int    `json:"k"`
	PAtLeastK  string `json:"pAtLeastK"`
	Multiplier string `json:"multiplier"`
}

type tippotCase struct {
	Name          string           `json:"name"`
	Probabilities []string         `json:"probabilities"`
	MarginBp      int              `json:"marginBp"`
	Tiers         []tippotCaseTier `json:"tiers"`
}

type fixture struct {
	Tiple  []tipleCase  `json:"tiple"`
	Tippot []tippotCase `json:"tippot"`
}

func loadFixture(t *testing.T) fixture {
	t.Helper()
	// services/settlement/internal/products → ../../../../docs/fixtures/products.json
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	path := filepath.Join(wd, "..", "..", "..", "..", "docs", "fixtures", "products.json")
	body, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture %s: %v", path, err)
	}
	var f fixture
	if err := json.Unmarshal(body, &f); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}
	return f
}

func parseProbs(t *testing.T, in []string) []float64 {
	t.Helper()
	out := make([]float64, len(in))
	for i, s := range in {
		p, err := ParseProbability(s)
		if err != nil {
			t.Fatalf("parse %s: %v", s, err)
		}
		out[i] = p
	}
	return out
}

func TestTipleGoldenFixture(t *testing.T) {
	f := loadFixture(t)
	for _, tc := range f.Tiple {
		t.Run(tc.Name, func(t *testing.T) {
			probs := parseProbs(t, tc.Probabilities)
			q, err := PriceTiple(probs, tc.MarginBp)
			if err != nil {
				t.Fatalf("price: %v", err)
			}
			if q.OfferedOdds != tc.OfferedOdds {
				t.Errorf("offeredOdds: got %s want %s", q.OfferedOdds, tc.OfferedOdds)
			}
			if math.Abs(q.FairProbability-tc.FairProbability) > 1e-9 {
				t.Errorf("fairProbability: got %v want %v", q.FairProbability, tc.FairProbability)
			}
		})
	}
}

func TestTippotGoldenFixture(t *testing.T) {
	f := loadFixture(t)
	for _, tc := range f.Tippot {
		t.Run(tc.Name, func(t *testing.T) {
			probs := parseProbs(t, tc.Probabilities)
			q, err := PriceTippot(probs, tc.MarginBp)
			if err != nil {
				t.Fatalf("price: %v", err)
			}
			if len(q.Tiers) != len(tc.Tiers) {
				t.Fatalf("tier count: got %d want %d", len(q.Tiers), len(tc.Tiers))
			}
			for i := range tc.Tiers {
				got, want := q.Tiers[i], tc.Tiers[i]
				if got.K != want.K {
					t.Errorf("tier[%d].k: got %d want %d", i, got.K, want.K)
				}
				if got.PAtLeastK != want.PAtLeastK {
					t.Errorf("tier[%d].pAtLeastK: got %s want %s", i, got.PAtLeastK, want.PAtLeastK)
				}
				if got.Multiplier != want.Multiplier {
					t.Errorf("tier[%d].multiplier: got %s want %s", i, got.Multiplier, want.Multiplier)
				}
			}
		})
	}
}

func TestTippotMultiplierStrictlyIncreasing(t *testing.T) {
	cases := []struct {
		probs []float64
		bp    int
	}{
		{[]float64{0.1, 0.2, 0.3, 0.4, 0.5}, 1500},
		{[]float64{0.5, 0.5, 0.5}, 1500},
		{[]float64{0.05, 0.1, 0.15, 0.2, 0.25, 0.3}, 0},
		{[]float64{0.9, 0.9, 0.9, 0.9}, 1500},
	}
	for _, c := range cases {
		q, err := PriceTippot(c.probs, c.bp)
		if err != nil {
			t.Fatalf("price: %v", err)
		}
		prev := math.Inf(-1)
		for _, tier := range q.Tiers {
			m, err := strconv.ParseFloat(tier.Multiplier, 64)
			if err != nil {
				t.Fatalf("parse %s: %v", tier.Multiplier, err)
			}
			if !(m > prev) {
				t.Errorf("multiplier should strictly increase: prev=%v m=%v", prev, m)
			}
			prev = m
		}
	}
}

func TestRejectsBadInputs(t *testing.T) {
	if _, err := PriceTiple([]float64{0.5, 1.0}, 1500); err == nil {
		t.Error("expected error for prob=1.0")
	}
	if _, err := PriceTiple([]float64{0.5}, 1500); err == nil {
		t.Error("expected error for too few legs")
	}
	if _, err := PriceTiple([]float64{0.5, 0.5}, -1); err == nil {
		t.Error("expected error for negative margin")
	}
	if _, err := PriceTippot([]float64{0.5, -0.1}, 1500); err == nil {
		t.Error("expected error for negative prob")
	}
}

File: packages/types/src/products.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { priceTiple, priceTippot, parseProbability } from "./products.js";

const here = dirname(fileURLToPath(import.meta.url));
const fixturePath = join(here, "..", "..", "..", "docs", "fixtures", "products.json");
const fixture = JSON.parse(readFileSync(fixturePath, "utf8")) as {
  tiple: Array<{
    name: string;
    probabilities: string[];
    marginBp: number;
    offeredOdds: string;
    fairProbability: number;
  }>;
  tippot: Array<{
    name: string;
    probabilities: string[];
    marginBp: number;
    tiers: Array<{ k: number; pAtLeastK: string; multiplier: string }>;
  }>;
};

test("tiple golden fixture", () => {
  for (const tc of fixture.tiple) {
    const probs = tc.probabilities.map(parseProbability);
    const q = priceTiple(probs, tc.marginBp);
    assert.equal(q.offeredOdds, tc.offeredOdds, `${tc.name}: offeredOdds`);
    // fairProbability is a float; allow tiny ULP diff but expect ~12 dp
    // agreement. The fixture stores it to 6-ish dp.
    const diff = Math.abs(q.fairProbability - tc.fairProbability);
    assert.ok(diff < 1e-9, `${tc.name}: fairProbability diff ${diff}`);
  }
});

test("tippot golden fixture", () => {
  for (const tc of fixture.tippot) {
    const probs = tc.probabilities.map(parseProbability);
    const q = priceTippot(probs, tc.marginBp);
    assert.equal(q.tiers.length, tc.tiers.length, `${tc.name}: tier count`);
    for (let i = 0; i < tc.tiers.length; i++) {
      const got = q.tiers[i]!;
      const want = tc.tiers[i]!;
      assert.equal(got.k, want.k, `${tc.name}: tier[${i}].k`);
      assert.equal(got.pAtLeastK, want.pAtLeastK, `${tc.name}: tier[${i}].pAtLeastK`);
      assert.equal(got.multiplier, want.multiplier, `${tc.name}: tier[${i}].multiplier`);
    }
  }
});

test("tippot multiplier is strictly increasing in k", () => {
  // Property: M_j > M_{j-1} for j ≥ 2 — invariant of the tier
  // construction. Catches algorithm regressions even if fixture rots.
  const cases: Array<{ probs: number[]; bp: number }> = [
    { probs: [0.1, 0.2, 0.3, 0.4, 0.5], bp: 1500 },
    { probs: [0.5, 0.5, 0.5], bp: 1500 },
    { probs: [0.05, 0.1, 0.15, 0.2, 0.25, 0.3], bp: 0 },
    { probs: [0.9, 0.9, 0.9, 0.9], bp: 1500 },
  ];
  for (const c of cases) {
    const q = priceTippot(c.probs, c.bp);
    let prev = -Infinity;
    for (const tier of q.tiers) {
      const m = Number(tier.multiplier);
      assert.ok(m > prev, `multiplier should strictly increase: prev=${prev} m=${m}`);
      prev = m;
    }
  }
});

test("tippot expected payout matches 1/(1+margin) up to floor truncation", () => {
  // Property: E[payout]/stake = 1/(1+m) exactly in real math. Floor
  // truncation makes the actual sum a hair lower. We require the gap
  // to be ≤ N × 10^-decimals (1 ULP per tier) — a tight bound that
  // catches any algorithmic drift.
  const decimals = 4;
  const ulp = Math.pow(10, -decimals);
  const cases: Array<{ probs: number[]; bp: number }> = [
    { probs: [0.5, 0.5, 0.5], bp: 1500 },
    { probs: [0.2, 0.2, 0.2, 0.2, 0.2], bp: 0 },
    { probs: [0.3, 0.4, 0.5, 0.6], bp: 1500 },
  ];
  for (const c of cases) {
    const q = priceTippot(c.probs, c.bp);
    const n = c.probs.length;
    // Reproduce the PMF inline to compute E[payout].
    let pmf = new Array(n + 1).fill(0);
    pmf[0] = 1;
    for (const p of c.probs) {
      const next = new Array(n + 1).fill(0);
      for (let k = 0; k <= pmf.length - 1; k++) {
        const stayLose = pmf[k] * (1 - p);
        const gainWin = k > 0 ? pmf[k - 1] * p : 0;
        next[k] = stayLose + gainWin;
      }
      pmf = next;
    }
    let expected = 0;
    for (let j = 1; j <= n; j++) {
      expected += pmf[j] * Number(q.tiers[j - 1]!.multiplier);
    }
    const target = 1 / (1 + c.bp / 10000);
    assert.ok(
      expected <= target + 1e-9,
      `payout exceeds margin target: expected=${expected} target=${target}`,
    );
    assert.ok(
      expected >= target - n * ulp,
      `payout under-pays bettor too much: expected=${expected} target=${target} (allowed slack ${n * ulp})`,
    );
  }
});

test("rejects out-of-range probabilities", () => {
  assert.throws(() => priceTiple([0.5, 1.0], 1500), /\(0,1\)/);
  assert.throws(() => priceTiple([0.5, 0], 1500), /\(0,1\)/);
  assert.throws(() => priceTippot([0.5, -0.1], 1500), /\(0,1\)/);
});

test("rejects too few legs", () => {
  assert.throws(() => priceTiple([0.5], 1500), /≥ 2 legs/);
  assert.throws(() => priceTippot([0.5], 1500), /≥ 2 legs/);
});

test("rejects out-of-range margin", () => {
  assert.throws(() => priceTiple([0.5, 0.5], -1), /margin_bp/);
  assert.throws(() => priceTiple([0.5, 0.5], 200001), /margin_bp/);
});
